GX-TimeFrequency/README.md

```diff
@@ -14,9 +11,96 @@ make -j
+(see the top-level README) change to `<GX_ROOT>/<BUILD_DIR>` and type `ctest`. 
+Additionally, one can change to the test folder and explicitly run the pytest 
+command from there:
+
+```bash
+cd <GX_ROOT>/<BUILD_DIR>/test/time-frequency
+pytest -s 
+# or
+pytest -s test_name.py
+```
+
+## Calling the MiniMax Library
+
+The minimax grid generation is called like so:
+
+```fortran
+use gx_minimax, only: gx_minimax_grid
+
+! Declarations
+integer :: n_mesh_points
+real(dp) :: e_transition_min, e_transition_max
+real(dp), allocatable :: tau_mesh(:), tau_weights(:)
+real(dp), allocatable :: freq_mesh(:), freq_weights(:)
+real(dp), allocatable :: cos_tau_to_freq_weights(:, :)
+real(dp), allocatable :: cos_freq_to_tau_weights(:, :)
+real(dp), allocatable :: sinft_tau_to_freq_weights(:, :)
+real(dp) :: max_errors(3)
+real(dp) :: cosft_duality_error
+integer :: ierr
+
+call gx_minimax_grid(n_mesh_points, e_transition_min, e_transition_max, &
+                     tau_mesh, tau_weights, &
+                     freq_mesh, freq_weights, &
+                     cos_tau_to_freq_weights, cos_freq_to_tau_weights, &
+                     sinft_tau_to_freq_weights, &
+                     max_errors, cosft_duality_error, ierr)
+```
+
+For a description of the variables, please consult `src/mp2_grids.F90`.
+For an example of how to call `gx_minimax_grid`, please consult `test/test_gx_minimax_grid.f90`.
+
+Additionally, one can also call a utility routine to query whether the 
+number of imaginary-time points has a corresponding grid tabulation:
+
+```fortran
+use api_utilites, only: gx_check_ntau, gx_get_error_message
+
+call gx_check_ntau(ntau, msg, ierr)
+```
+
+## For Developers
+
+### Adding a New Test
+
+Write a `test_name.py` and `test_name.f90` in `test/`. 
+`test_name.f90` should accept some input parameter/s, make a library call and
+write the result to file. `test_name.py` should provide the input parameters
+(the simplest way is to write to file), execute the binary built from 
+`test_name.f90`, parse the result and make the assertions. This imples the
+reference results should also be stored in `test_name.py`.
+
+`test_name.py` *must* define the binary name, by convention `test_name.exe`
+and define the root path in which to look for the binary. This can be achieved
+with:
 
-Application tests are run with the pytest framework. From the GreenX root directory, build the
-code then type:
+```python
+root = Path(__file__).parent.parent.parent
+```
+  
+where the level of nesting depends upon how one defines the `test/` directory
+structure in the `<BUILD>` folder. See `developers.md` for a discussion of how
+one can (and should) make this more robust. 
+
+Additionally, one should ensure that both python and fortran read/write to `tmp_path`.
+The `tmp_path` fixture will provide a temporary directory unique to the (py)test 
+invocation, created in the base temporary directory.
```

Comment:
Can you elaborate? One does not want to send everything to `tmp_path` because?
I precisely used it because I want a common path, with will automatically get cleaned up.

Reply:
If we *do* want to send everything to tmp_path, then we should automate and enforce that, rather than tell the user to "remember" it.  (Although we may still want to tell the user about that.)

```
@pytest.fixture(autouse=True)
def testdir(tmp_path):
   oldpath = os.cwd()
   os.chdir(tmp_path)
   yield
   os.chdir(oldpath)
   print(f'Testdir: {path}')
```

(The final print statement makes sure that the user can actually see *where* a test failed.)

Reply:
With that fixture, any test is free to leave as much garbage as it wants

Reply:
To clarify, one *normally* wants most tests to *not* use the filesystem, but maybe in our case it makes sense.

Reply:
And where does one define this fixture? Also in the conftest?

> With that fixture, any test is free to leave as much garbage as it wants

Is this not a bad thing? Don't we want automated clean-up?

> To clarify, one normally wants most tests to not use the filesystem

Agreed in the context of unit tests. If we had structured input/output, a fortran unit test framework in place, or python bindings for the lib, this is avoidable. I don't foresee this happening though, it will be a case of whatever the first working version is, is also the final state.

Reply:
You said

> Can you elaborate? One does not want to send everything to tmp_path because?

....because it's a bit messy, it is cleaner if tests that need files are explicit about that requirement by depending on tmp_path.

But you also said

> Is this not a bad thing? Don't we want automated clean-up?

Well yes, as noted above, it is not so clean.  You decide what you want. :)

Anyway automated cleanup is easy: tmp_path goes under /tmp so the system will delete it e.g. at reboot or whenever it feels like it.  Also, pytest always checks for files from older pytest runs, so at most a few old runs' worth of tempfiles will be kept at any time.  So it's already as automatic as it needs to be and we don't need to care.

Reply:
> And where does one define this fixture? Also in the conftest?

Any fixture needed in many places goes in conftest.

Reply:
Is your suggestion slightly different in that this is doing the work in tmp_path rather than sending files to/reading files from tmp_path?

Ignorant question, how does yield work in this context?